### ISSUE TITLE
Navigate to custom Routes like DialogRoute<T>

### ISSUE THREAD

**Issue by hauketoenjes** [lulupointu/vrouter#31]
I am currently working on a project where im using this package for navigation. It works absolutely fine for all that i have been doing with it so far 👍🏼 . 

But i am currently trying to add a custom route to the navigation stack because im working a lot with dialogs in the App. 

Example: 
There is a page that shows a list with let's say books. If you click on a book it opens a dialog to show more information about the book. 

What i want to achieve is pushing this Dialog with `context.vrouter.push('/books/1/detail')` instead of `showDialog()` and a button (which does not allow for deep-links with a dialog). Do you think that this is a feature that is possible and useful to implement? I am thinking of a way to give VRouter a Custom Route like `DialogRoute<T>` to navigate to.  

From what i understand currently the package decides, based on the platform, to use `MaterialPageRoute<T>` or `CupertinoPageRoute<T>` only.

**Comment by lulupointu** [comment]
Could you write down a small 1 page snippet using `showDialog()` accomplishing what you want to do ?
Like, 1 page with a list of book + `onPressed` shows the dialog.

I think this is actually possible to implement with the current VRouter but I want to be sure that I understand what you are trying to do.

**Comment by hauketoenjes** [comment]
Hi, thanks for the fast reply. Here's a DartPad of what im trying to accomplish: 

https://dartpad.dev/6e0c3d919f6415ad39284e2dbebf96be

The GitHub gist to this DartPad can be found under: 
https://gist.github.com/hauketoenjes/6e0c3d919f6415ad39284e2dbebf96be

The interesting part of the code is the following: 
```dart
          ListTile(
            title: Text('Book 1'),
            subtitle: Text('An interesting Book'),
            onTap: () {
              // Here i want to use context.vrouter.push('/books/:id/detail') instead of using
              // showDialog. Declaring it as route would make it possible to access this Dialog
              // without the onTap() Method when let's say sharing a deep Link to this book. The user
              // would be able to see the dialog directly and the navigation stack behind it.
              showDialog(
                context: context,
                builder: (BuildContext context) {
                  return AlertDialog(
                    title: Text('Book 1 Information'),
                    content: Text('More Information'),
                  );
                },
              );
            },
          ),
```

**Comment by lulupointu** [comment]
Amazing, you just made me realize something which is possible with VRouter that I had no idea! Thanks!

Here is the beauty:

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

void main() {
  runApp(MyApp());
}

class Book {
  final String title;
  final String description;

  Book({required this.title, required this.description});
}

class MyApp extends StatelessWidget {
  final books = [
    Book(title: 'Book1', description: 'A first interesting book'),
    Book(title: 'Book', description: 'A second interesting book'),
  ];

  @override
  Widget build(BuildContext context) {
    return VRouter(
      initialUrl: '/books',
      routes: [
        VWidget(
          path: '/books',
          widget: BooksScreen(books: books),
          stackedRoutes: [
            VPage(
              path: ':bookId/details',
              pageBuilder: (key, child) => DialogPage(key: key, child: child),
              widget: Builder(
                builder: (context) => BookAlertDialog(
                  book: books[int.parse(context.vRouter.pathParameters['bookId']!)],
                ),
              ),
            ),
          ],
        ),
      ],
    );
  }
}

class BooksScreen extends StatelessWidget {
  final List<Book> books;

  const BooksScreen({required this.books});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Books")),
      body: ListView(
        children: [
          for (final book in books)
            ListTile(
              title: Text(book.title),
              subtitle: Text(book.description),
              onTap: () => context.vRouter.push('/books/${books.indexOf(book)}/details'),
            ),
        ],
      ),
    );
  }
}

class BookAlertDialog extends StatelessWidget {
  final Book book;

  const BookAlertDialog({required this.book});

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: Text(book.title),
      content: Text(book.description),
    );
  }
}

class DialogPage<T> extends Page<T> {
  @override
  final LocalKey key;

  final Widget child;

  DialogPage({
    required this.key,
    required this.child,
  });

  @override
  Route<T> createRoute(BuildContext context) {
    return DialogRoute<T>(
      context: context,
      builder: (_) => child,
      settings: this,
    );
  }
}
```

The trick here is that `VRouter` allows you to pass any page, so any route. You just have to create a custom page and build a `DialogRoute`. This is basically what `showDialog` does really, nothing fancy and pretty elegant I would say :rocket:

**Comment by hauketoenjes** [comment]
Hi, 
thanks for the example. That ist what i was looking for. 👍🏼 

Do you think it is a Flutter Bug that this approach doesn't work when directly accessing the url through the browser? It works when navigating in the app and when changing the url in the browsers bar while the app is "running". 

It doesn't work, when you run the app in release mode as let's say web server and you access the url `[...]/books/1/details` directly. After loading you are redirected to the `/books` page

**Comment by lulupointu** [comment]
Oh this was a bug with `initialUrl` breaking deep-liking, this is now fixed in v1.1.0+15.

**Comment by hauketoenjes** [comment]
Thanks for the quick update, works like a charm now!
